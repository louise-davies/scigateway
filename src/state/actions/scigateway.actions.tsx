import axios from 'axios';
import { Action, AnyAction } from 'redux';
import log from 'loglevel';
import {
  ConfigureStringsType,
  ConfigureStringsPayload,
  ToggleDrawerType,
  AuthFailureType,
  AuthSuccessType,
  ApplicationStrings,
  SignOutType,
  FeatureSwitchesPayload,
  ConfigureFeatureSwitchesType,
  FeatureSwitches,
  LoadingAuthType,
  LoadedAuthType,
  DismissNotificationType,
  DismissNotificationPayload,
  RequestPluginRerenderType,
  AuthProviderPayload,
  LoadAuthProviderType,
  SiteLoadingType,
  SiteLoadingPayload,
  ConfigureAnalyticsPayload,
  ConfigureAnalyticsType,
  InitialiseAnalyticsType,
  ToggleHelpType,
  AddHelpTourStepsType,
  AddHelpTourStepsPayload,
  InvalidateTokenType,
  SendThemeOptionsType,
  SendThemeOptionsPayload,
  LoadDarkModePreferenceType,
  LoadDarkModePreferencePayload,
  StartUrlPayload,
  RegisterStartUrlType,
  ScheduledMaintenanceState,
  LoadScheduledMaintenanceStateType,
  ScheduledMaintenanceStatePayLoad,
  MaintenanceStatePayLoad,
  LoadMaintenanceStateType,
  MaintenanceState,
  NotificationType,
} from '../scigateway.types';
import { ActionType, ThunkResult, StateType } from '../state.types';
import loadMicroFrontends from './loadMicroFrontends';
import { push } from 'connected-react-router';
import { ThunkAction } from 'redux-thunk';
import { Step } from 'react-joyride';
import { Theme } from '@material-ui/core/styles/createMuiTheme';

export const configureStrings = (
  appStrings: ApplicationStrings
): ActionType<ConfigureStringsPayload> => ({
  type: ConfigureStringsType,
  payload: {
    res: appStrings,
  },
});

export const loadStrings = (path: string): ThunkResult<Promise<void>> => {
  return async (dispatch) => {
    await axios
      .get(path)
      .then((res) => {
        dispatch(configureStrings(res.data));
      })
      .catch((error) =>
        log.error(`Failed to read strings from ${path}: ${error}`)
      );
  };
};

export const loadFeatureSwitches = (
  featureSwitches: FeatureSwitches
): ActionType<FeatureSwitchesPayload> => ({
  type: ConfigureFeatureSwitchesType,
  payload: {
    switches: featureSwitches,
  },
});

export const registerStartUrl = (
  startUrl: string
): ActionType<StartUrlPayload> => ({
  type: RegisterStartUrlType,
  payload: {
    startUrl: startUrl,
  },
});

export const addHelpTourSteps = (
  steps: Step[]
): ActionType<AddHelpTourStepsPayload> => ({
  type: AddHelpTourStepsType,
  payload: {
    steps,
  },
});

export const loadAuthProvider = (
  authProvider: string,
  authUrl?: string
): ActionType<AuthProviderPayload> => ({
  type: LoadAuthProviderType,
  payload: {
    authProvider,
    authUrl,
  },
});

export const loadScheduledMaintenanceState = (
  scheduledMaintenance: ScheduledMaintenanceState
): ActionType<ScheduledMaintenanceStatePayLoad> => ({
  type: LoadScheduledMaintenanceStateType,
  payload: {
    scheduledMaintenance: scheduledMaintenance,
  },
});

export const loadMaintenanceState = (
  maintenance: MaintenanceState
): ActionType<MaintenanceStatePayLoad> => ({
  type: LoadMaintenanceStateType,
  payload: {
    maintenance: maintenance,
  },
});

export const loadingAuthentication = (): Action => ({
  type: LoadingAuthType,
});

export const loadedAuthentication = (): Action => ({
  type: LoadedAuthType,
});

export const unauthorised = (): Action => ({
  type: AuthFailureType,
});

export const authorised = (): Action => ({
  type: AuthSuccessType,
});

export const invalidToken = (): Action => ({
  type: InvalidateTokenType,
});

export const siteLoadingUpdate = (
  loading: boolean
): ActionType<SiteLoadingPayload> => ({
  type: SiteLoadingType,
  payload: {
    loading,
  },
});

export const configureAnalytics = (
  id: string
): ActionType<ConfigureAnalyticsPayload> => ({
  type: ConfigureAnalyticsType,
  payload: {
    id,
  },
});

export const initialiseAnalytics = (): Action => ({
  type: InitialiseAnalyticsType,
});

export const configureSite = (): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    await axios
      .get(`/settings.json`)
      .then((res) => {
        const settings = res.data;

        // invalid settings.json - Use JSON.parse to give detailed error info
        if (typeof settings !== 'object') {
          throw Error('Invalid format');
        }

        if (settings['ga-tracking-id']) {
          dispatch(configureAnalytics(settings['ga-tracking-id']));
        }

        dispatch(
          loadAuthProvider(settings['auth-provider'], settings['authUrl'])
        );

        const loadingPromises = [];

        const provider = getState().scigateway.authorisation.provider;

        if (provider.fetchMaintenanceState) {
          provider.fetchMaintenanceState().then((maintenanceState) => {
            dispatch(loadMaintenanceState(maintenanceState));
          });
        }

        // after auth provider is set then the token needs to be verified
        // also attempt to auto login if the auth provider allows it
        if (provider.isLoggedIn()) {
          const verifyingLogin = provider
            .verifyLogIn()
            .then(() => {
              dispatch(authorised());
            })
            .catch(() => {
              if (provider.autoLogin) {
                dispatch(loadingAuthentication());
                loadingPromises.push(
                  provider
                    .autoLogin()
                    .then(() => {
                      dispatch(authorised());
                    })
                    .catch(() => {
                      dispatch(invalidToken());
                    })
                );
              } else {
                dispatch(invalidToken());
              }
            });

          loadingPromises.push(verifyingLogin);
        } else if (provider.autoLogin) {
          dispatch(loadingAuthentication());
          loadingPromises.push(
            provider
              .autoLogin()
              .then(() => {
                dispatch(authorised());
              })
              .catch(() => {
                dispatch(loadedAuthentication());
              })
          );
        }

        if (settings['features']) {
          dispatch(loadFeatureSwitches(settings['features']));
        }

        dispatch(addHelpTourSteps(settings['help-tour-steps']));

        if (settings['startUrl']) {
          dispatch(registerStartUrl(settings['startUrl']));
        }

        const uiStringResourcesPath = !settings['ui-strings'].startsWith('/')
          ? '/' + settings['ui-strings']
          : settings['ui-strings'];
        const loadingResources = dispatch(loadStrings(uiStringResourcesPath));
        loadingPromises.push(loadingResources);

        const loadingPlugins = loadMicroFrontends.init(settings.plugins);
        loadingPromises.push(loadingPlugins);

        return Promise.all(loadingPromises).then(() => {
          dispatch(siteLoadingUpdate(false));
        });
      })
      .catch((error) => {
        log.error(`Error loading settings.json: ${error.message}`);
      });

    // load dark mode preference from local storage into store
    // otherwise, fetch system preference
    const darkModeLocalStorage = localStorage.getItem('darkMode');
    if (darkModeLocalStorage) {
      const preference = darkModeLocalStorage === 'true' ? true : false;
      dispatch(loadDarkModePreference(preference));
    } else {
      const mq = window.matchMedia('(prefers-color-scheme: dark)');
      if (mq) dispatch(loadDarkModePreference(mq.matches));
    }

    const provider = getState().scigateway.authorisation.provider;
    if (provider.fetchScheduledMaintenanceState) {
      provider
        .fetchScheduledMaintenanceState()
        .then((scheduledMaintenanceState) => {
          dispatch(loadScheduledMaintenanceState(scheduledMaintenanceState));

          // Checking the state in the GET response because it does not get
          // loaded into the store before this check is performed
          if (scheduledMaintenanceState['show']) {
            displayScheduledMaintenanceBanner(
              scheduledMaintenanceState['message']
            );
          }
        });
    }
  };
};

const displayScheduledMaintenanceBanner = (message: string): void => {
  document.dispatchEvent(
    new CustomEvent('scigateway', {
      detail: {
        type: NotificationType,
        payload: {
          severity: 'warning',
          message: message,
        },
      },
    })
  );
};

export const toggleDrawer = (): Action => ({
  type: ToggleDrawerType,
});

export const toggleHelp = (): Action => ({
  type: ToggleHelpType,
});

export const signOut = (): ThunkAction<void, StateType, null, AnyAction> => (
  dispatch
) => {
  dispatch({ type: SignOutType });
  dispatch(push('/'));
};

export const verifyUsernameAndPassword = (
  username: string,
  password: string
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    // will be replaced with call to login API for authentification
    dispatch(loadingAuthentication());
    const authProvider = getState().scigateway.authorisation.provider;
    await authProvider
      .logIn(username, password)
      .then(() => {
        dispatch(authorised());

        // redirect the user to the original page they were trying to get to
        // the referrer is added by the redirect in routing.component.tsx
        const previousRouteState = getState().router.location.state;
        dispatch(
          push(
            previousRouteState && previousRouteState.referrer
              ? previousRouteState.referrer
              : '/'
          )
        );
      })
      .catch(() => {
        // probably want to do something smarter with
        // err.response.status (e.g. 403 or 500)
        dispatch(unauthorised());
      });
  };
};

export const requestPluginRerender = (): ActionType<{
  broadcast: boolean;
}> => ({
  type: RequestPluginRerenderType,
  payload: {
    broadcast: true,
  },
});

export const sendThemeOptions = (
  theme: Theme
): ActionType<SendThemeOptionsPayload> => ({
  type: SendThemeOptionsType,
  payload: {
    theme,
    broadcast: true,
  },
});

export const loadDarkModePreference = (
  darkMode: boolean
): ActionType<LoadDarkModePreferencePayload> => ({
  type: LoadDarkModePreferenceType,
  payload: {
    darkMode: darkMode,
  },
});

export const dismissMenuItem = (
  index: number
): ActionType<DismissNotificationPayload> => {
  return {
    type: DismissNotificationType,
    payload: {
      index,
    },
  };
};
