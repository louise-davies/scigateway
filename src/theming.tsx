import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
import { ThemeOptions, Theme } from '@material-ui/core/styles/createMuiTheme';
import 'react-redux-toastr/lib/css/react-redux-toastr.min.css';
import React from 'react';
import { StateType } from './state/state.types';
import { connect, useSelector } from 'react-redux';

export interface UKRIThemeOptions extends ThemeOptions {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

export interface UKRITheme extends Theme {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

/* Colours that may be used across light/dark modes e.g. the main app bar */
const STATIC_COLOURS = {
  darkBlue: '#003088',
  orange: '#FF6900',
};

/* Main colours used for dark/light modes respectively */
interface ThemeColours {
  primary: string;
  secondary: string;
  background: string;
  paper: string;
  blue: string;
  orange: string;
  red: string;
  grey: string;
}

const DARK_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

//For experimenting
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const DARK_MODE_COLOURS_HIGH_CONTRAST: ThemeColours = {
  primary: '#86B4FF',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

const LIGHT_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#003088',
  background: '#FAFAFA',
  paper: '#FFF',
  blue: '#003088',
  orange: '#C34F00',
  red: '#AC1600',
  grey: '#727272',
};

export const buildTheme = (darkModePreference: boolean): Theme => {
  let options: UKRIThemeOptions;
  const colours = darkModePreference ? DARK_MODE_COLOURS : LIGHT_MODE_COLOURS;

  const overrides = {
    MuiLink: {
      root: {
        color: colours.blue,
      },
    },
    MuiTabs: {
      indicator: {
        color: STATIC_COLOURS.darkBlue,
        textDecoration: 'underline',
      },
    },
    MuiFormLabel: {
      root: {
        '&$error': {
          color: colours.red,
        },
        '&$focused': {
          color: colours.blue,
        },
      },
      asterisk: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiBadge: {
      colorPrimary: {
        backgroundColor: STATIC_COLOURS.orange,
      },
    },
    MuiInput: {
      underline: {
        '&$error:after': {
          borderBottomColor: colours.red,
        },
        '&:after': {
          borderBottomColor: colours.blue,
        },
      },
    },
    MuiOutlinedInput: {
      root: {
        '&$error $notchedOutline': {
          borderColor: colours.red,
        },
      },
    },
    MuiFormHelperText: {
      root: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiPickersToolbar: {
      toolbar: {
        backgroundColor: STATIC_COLOURS.darkBlue,
      },
    },
    MuiPickersCalendarHeader: {
      dayLabel: {
        color: colours.grey,
      },
    },
    MuiPickersDay: {
      current: {
        color: colours.blue,
      },
      dayDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersYear: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      yearSelected: {
        color: colours.blue,
      },
      yearDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersMonth: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      monthSelected: {
        color: colours.blue,
      },
      monthDisabled: {
        color: colours.grey,
      },
    },
  };

  if (darkModePreference) {
    options = {
      palette: {
        // Light/dark mode
        type: 'dark',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#257fff',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  } else {
    options = {
      palette: {
        // Light/dark mode
        type: 'light',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#1E5DF8',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  }

  return createMuiTheme(options);
};

function mapThemeProviderStateToProps(
  state: StateType
): { prefersDarkMode: boolean } {
  return {
    prefersDarkMode: state.scigateway.darkMode,
  };
}

const SciGatewayThemeProvider = (props: {
  children: React.ReactNode;
}): React.ReactElement<{
  children: React.ReactNode;
}> => {
  const darkModePreference: boolean = useSelector(
    (state: StateType) => state.scigateway.darkMode
  );
  return (
    <MuiThemeProvider theme={buildTheme(darkModePreference)}>
      {props.children}
    </MuiThemeProvider>
  );
};

export const ConnectedThemeProvider = connect(mapThemeProviderStateToProps)(
  SciGatewayThemeProvider
);
