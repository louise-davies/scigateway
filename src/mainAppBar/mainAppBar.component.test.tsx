import React from 'react';
import MainAppBarComponent from './mainAppBar.component';
import { createLocation } from 'history';
import { StateType } from '../state/state.types';
import { PluginConfig } from '../state/scigateway.types';
import configureStore from 'redux-mock-store';
import { push } from 'connected-react-router';
import { initialState } from '../state/reducers/scigateway.reducer';
import {
  loadHighContrastModePreference,
  toggleDrawer,
  toggleHelp,
} from '../state/actions/scigateway.actions';
import { Provider } from 'react-redux';
import TestAuthProvider from '../authentication/testAuthProvider';
import { buildTheme } from '../theming';
import { ThemeProvider, StyledEngineProvider } from '@mui/material/styles';
import { loadDarkModePreference } from '../state/actions/scigateway.actions';
import { createMemoryHistory, History } from 'history';
import { ReactWrapper, mount } from 'enzyme';
import { Router } from 'react-router-dom';
import ScigatewayLogo from '../images/scigateway-white-text-blue-mark-logo.svg';

describe('Main app bar component', () => {
  let mockStore;
  let state: StateType;
  let history: History;

  const createWrapper = (testStore): ReactWrapper => {
    return mount(
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={theme}>
          <Provider store={testStore}>
            <Router history={history}>
              <MainAppBarComponent />
            </Router>
          </Provider>
        </ThemeProvider>
      </StyledEngineProvider>
    );
  };

  beforeEach(() => {
    history = createMemoryHistory();

    mockStore = configureStore();
    state = {
      scigateway: {
        ...initialState,
        features: { ...initialState.features, showHelpPageButton: true },
      },
      router: { location: createLocation('/') },
    };
    state.scigateway.authorisation.provider = new TestAuthProvider('token123');
  });

  const theme = buildTheme(false);

  it('app bar renders correctly', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);
    expect(wrapper.find('MainAppBar').props()).toMatchSnapshot();
  });

  it('does not render Help button when feature is false', () => {
    state.scigateway.features.showHelpPageButton = false;
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);
    expect(wrapper.find('MainAppBar').props()).toMatchSnapshot();
  });

  it('uses single plugin logo when feature is true', () => {
    state.scigateway.features.singlePluginLogo = true;
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);
    expect(wrapper.find('MainAppBar').props()).toMatchSnapshot();
  });

  it('app bar indented when drawer is open', () => {
    state.scigateway.drawerOpen = true;
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);
    expect(wrapper.find('MainAppBar').props()).toMatchSnapshot();
  });

  it('sends toggleDrawer action when site has loaded', () => {
    state.scigateway.drawerOpen = false;
    state.scigateway.siteLoading = false;
    const testStore = mockStore(state);
    createWrapper(testStore);

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleDrawer());
  });

  it('sends toggleDrawer action when menu clicked (open drawer)', () => {
    state.scigateway.drawerOpen = true;
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    wrapper.find('button').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleDrawer());
  });

  it('sends toggleDrawer action when menu clicked (close drawer)', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    wrapper.find('button').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleDrawer());
  });

  it('redirects to base url when title clicked', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    wrapper.find('button[aria-label="home-page"]').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(push('/'));
  });

  it('redirects to Help page when Help button clicked', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    wrapper.find('button[aria-label="help-page"]').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(push('/help'));
  });

  it('redirects to Admin page when Admin button clicked (maintenance is default)', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    wrapper.find('button[aria-label="admin-page"]').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(push('/admin/maintenance'));
  });

  it('redirects to Admin page when Admin button clicked (download is default)', () => {
    state.scigateway.adminPageDefaultTab = 'download';
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    wrapper.find('button[aria-label="admin-page"]').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(push('/admin/download'));
  });

  it('sends toggleHelp action when help button is clicked', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    wrapper.find('button[aria-label="help"]').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleHelp());
  });

  it('opens settings when button clicked', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    expect(wrapper.find('#settings').first().prop('open')).toBeFalsy();

    wrapper
      .find('button[aria-label="open-browser-settings"]')
      .simulate('click');

    expect(wrapper.find('#settings').first().prop('open')).toBeTruthy();

    expect(wrapper.find('#item-dark-mode').first().text()).toEqual(
      'switch-dark-mode'
    );
    expect(wrapper.find('#item-high-contrast-mode').first().text()).toEqual(
      'switch-high-contrast-on'
    );
  });

  it('settings display correctly when dark and high contrast modes are enabled', () => {
    state.scigateway.darkMode = true;
    state.scigateway.highContrastMode = true;
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    expect(wrapper.find('#settings').first().prop('open')).toBeFalsy();

    wrapper
      .find('button[aria-label="open-browser-settings"]')
      .simulate('click');

    expect(wrapper.find('#settings').first().prop('open')).toBeTruthy();

    expect(wrapper.find('#item-dark-mode').first().text()).toEqual(
      'switch-light-mode'
    );
    expect(wrapper.find('#item-high-contrast-mode').first().text()).toEqual(
      'switch-high-contrast-off'
    );
  });

  it('opens cookie policy/management page if manage cookies clicked', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    // Click the user menu button and click on the manage cookies menu item.
    wrapper
      .find('button[aria-label="open-browser-settings"]')
      .simulate('click');
    wrapper.find('#item-manage-cookies').last().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(push('/cookies'));
  });

  it('sends load dark mode prefrence action if toggle dark mode is clicked', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    // Click the user menu button and click on the manage cookies menu item.
    wrapper
      .find('button[aria-label="open-browser-settings"]')
      .simulate('click');
    wrapper.find('#item-dark-mode').last().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(loadDarkModePreference(true));
  });

  it('sends load high contrast mode prefrence action if toggle high contrast mode is clicked', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    // Click the user menu button and click on the manage cookies menu item.
    wrapper
      .find('button[aria-label="open-browser-settings"]')
      .simulate('click');
    wrapper.find('#item-high-contrast-mode').last().simulate('click');
    wrapper.update();

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(
      loadHighContrastModePreference(true)
    );
  });

  it('sets plugin logo', () => {
    const plugin: PluginConfig = {
      section: 'section',
      link: '/link',
      plugin: 'plugin',
      displayName: 'pluginName',
      order: 1,
      logoDarkMode: 'pluginLogo',
    };
    state.scigateway.plugins = [plugin];
    state.scigateway.siteLoading = false;

    const testStore = mockStore(state);
    // Need to attachTo something to ensure document.getElementById works as expected
    // https://stackoverflow.com/questions/43694975/jest-enzyme-using-mount-document-getelementbyid-returns-null-on-componen
    const holder = document.createElement('div');
    document.body.appendChild(holder);
    const wrapper = mount(
      <div id="plugin">
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <Provider store={testStore}>
              <Router history={history}>
                <MainAppBarComponent />
              </Router>
            </Provider>
          </ThemeProvider>
        </StyledEngineProvider>
      </div>,
      { attachTo: holder }
    );

    expect(wrapper.find('img').prop('src')).toEqual('pluginLogo');
  });

  it('sets Scigateway logo if no plugins are loaded or no plugins match found plugin or if plugin does not provide logo', () => {
    let testStore = mockStore(state);
    let wrapper = createWrapper(testStore);
    expect(wrapper.find('img').prop('src')).toEqual(ScigatewayLogo);

    let plugin: PluginConfig = {
      section: 'section',
      link: '/link',
      plugin: 'plugin1',
      displayName: 'pluginName',
      order: 1,
      logoDarkMode: 'pluginLogo',
    };
    state.scigateway.plugins = [plugin];

    testStore = mockStore(state);
    wrapper = mount(
      <div id="plugin2">
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <Provider store={testStore}>
              <Router history={history}>
                <MainAppBarComponent />
              </Router>
            </Provider>
          </ThemeProvider>
        </StyledEngineProvider>
      </div>
    );

    expect(wrapper.find('img').prop('src')).toEqual(ScigatewayLogo);

    plugin = {
      section: 'section',
      link: '/link',
      plugin: 'plugin',
      displayName: 'pluginName',
      order: 1,
    };
    state.scigateway.plugins = [plugin];

    testStore = mockStore(state);
    wrapper = mount(
      <div id="plugin">
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <Provider store={testStore}>
              <Router history={history}>
                <MainAppBarComponent />
              </Router>
            </Provider>
          </ThemeProvider>
        </StyledEngineProvider>
      </div>
    );

    expect(wrapper.find('img').prop('src')).toEqual(ScigatewayLogo);
  });

  it('sets first plugin logo when the singlePluginLogo setting is true', () => {
    let testStore = mockStore(state);
    let wrapper = createWrapper(testStore);
    expect(wrapper.find('img').prop('src')).toEqual(ScigatewayLogo);

    state.scigateway.features.singlePluginLogo = true;
    state.scigateway.plugins = [
      {
        section: 'section',
        link: '/link',
        plugin: 'plugin1',
        displayName: 'pluginName',
        order: 1,
        logoDarkMode: 'pluginLogo',
      },
      {
        section: 'section',
        link: '/link',
        plugin: 'plugin2',
        displayName: 'pluginName',
        order: 2,
        logoDarkMode: 'pluginLogo2',
      },
    ];
    state.scigateway.siteLoading = false;

    testStore = mockStore(state);
    wrapper = mount(
      <div id="plugin2">
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <Provider store={testStore}>
              <Router history={history}>
                <MainAppBarComponent />
              </Router>
            </Provider>
          </ThemeProvider>
        </StyledEngineProvider>
      </div>
    );

    expect(wrapper.find('img').prop('src')).toEqual('pluginLogo');
  });

  it('opens no notifications message if alert icon clicked and there are no alerts', () => {
    const testStore = mockStore(state);
    const wrapper = createWrapper(testStore);

    //Check notification only opens when clicked
    expect(
      wrapper.find('[aria-label="No notifications message"]').exists()
    ).toBeFalsy();

    wrapper
      .find('button[aria-label="Open notification menu"]')
      .simulate('click');

    expect(
      wrapper.find('[aria-label="No notifications message"]').exists()
    ).toBeTruthy();
  });
});
