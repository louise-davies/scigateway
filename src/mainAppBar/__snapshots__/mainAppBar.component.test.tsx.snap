// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main app bar component app bar indented when drawer is open 1`] = `
Object {
  "adminPageDefaultTab": undefined,
  "darkMode": false,
  "drawerOpen": true,
  "highContrastMode": false,
  "homepageUrl": undefined,
  "loading": true,
  "loggedIn": true,
  "logo": undefined,
  "manageCookies": [Function],
  "navigateToAdminPage": [Function],
  "navigateToHelpPage": [Function],
  "navigateToHome": [Function],
  "pathname": "/",
  "plugins": Array [],
  "res": undefined,
  "showAdminPageButton": true,
  "showHelpPageButton": true,
  "singlePluginLogo": false,
  "toggleDarkMode": [Function],
  "toggleDrawer": [Function],
  "toggleHelp": [Function],
  "toggleHighContrastMode": [Function],
}
`;

exports[`Main app bar component app bar renders correctly 1`] = `
Object {
  "adminPageDefaultTab": undefined,
  "darkMode": false,
  "drawerOpen": false,
  "highContrastMode": false,
  "homepageUrl": undefined,
  "loading": true,
  "loggedIn": true,
  "logo": undefined,
  "manageCookies": [Function],
  "navigateToAdminPage": [Function],
  "navigateToHelpPage": [Function],
  "navigateToHome": [Function],
  "pathname": "/",
  "plugins": Array [],
  "res": undefined,
  "showAdminPageButton": true,
  "showHelpPageButton": true,
  "singlePluginLogo": false,
  "toggleDarkMode": [Function],
  "toggleDrawer": [Function],
  "toggleHelp": [Function],
  "toggleHighContrastMode": [Function],
}
`;

exports[`Main app bar component does not render Help button when feature is false 1`] = `
Object {
  "adminPageDefaultTab": undefined,
  "darkMode": false,
  "drawerOpen": false,
  "highContrastMode": false,
  "homepageUrl": undefined,
  "loading": true,
  "loggedIn": true,
  "logo": undefined,
  "manageCookies": [Function],
  "navigateToAdminPage": [Function],
  "navigateToHelpPage": [Function],
  "navigateToHome": [Function],
  "pathname": "/",
  "plugins": Array [],
  "res": undefined,
  "showAdminPageButton": true,
  "showHelpPageButton": false,
  "singlePluginLogo": false,
  "toggleDarkMode": [Function],
  "toggleDrawer": [Function],
  "toggleHelp": [Function],
  "toggleHighContrastMode": [Function],
}
`;

exports[`Main app bar component uses single plugin logo when feature is true 1`] = `
Object {
  "adminPageDefaultTab": undefined,
  "darkMode": false,
  "drawerOpen": false,
  "highContrastMode": false,
  "homepageUrl": undefined,
  "loading": true,
  "loggedIn": true,
  "logo": undefined,
  "manageCookies": [Function],
  "navigateToAdminPage": [Function],
  "navigateToHelpPage": [Function],
  "navigateToHome": [Function],
  "pathname": "/",
  "plugins": Array [],
  "res": undefined,
  "showAdminPageButton": true,
  "showHelpPageButton": true,
  "singlePluginLogo": true,
  "toggleDarkMode": [Function],
  "toggleDrawer": [Function],
  "toggleHelp": [Function],
  "toggleHighContrastMode": [Function],
}
`;
